lib/HLSL/DxilLegalizeSampleOffsetPass.cpp

```diff
@@ -68,94 +70,151 @@ class DxilLegalizeSampleOffsetPass : public FunctionPass {
     TryUnrollLoop(illegalOffsets, F);
 
     // Collect offset again after mem2reg.
-    std::vector<Instruction *> ssaIllegalOffsets;
+    std::vector<std::pair<Value *, CallInst*>> ssaIllegalOffsets;
     CollectIllegalOffsets(ssaIllegalOffsets, F, hlslOP);
 
     // Run simple optimization to legalize offsets.
     LegalizeOffsets(ssaIllegalOffsets);
 
-    FinalCheck(illegalOffsets, F, hlslOP);
+    FinalCheck(F, hlslOP);
 
     return true;
   }
 
 private:
-  void TryUnrollLoop(std::vector<Instruction *> &illegalOffsets, Function &F);
-  void CollectIllegalOffsets(std::vector<Instruction *> &illegalOffsets,
+  void TryUnrollLoop(std::vector<std::pair<Value *, CallInst*>> &illegalOffsets, Function &F);
+  void CollectIllegalOffsets(std::vector<std::pair<Value *, CallInst*>> &illegalOffsets,
                              Function &F, hlsl::OP *hlslOP);
-  void CollectIllegalOffsets(std::vector<Instruction *> &illegalOffsets,
+  void CollectIllegalOffsets(std::vector<std::pair<Value *, CallInst*>> &illegalOffsets,
                              Function &F, DXIL::OpCode opcode,
                              hlsl::OP *hlslOP);
-  void LegalizeOffsets(const std::vector<Instruction *> &illegalOffsets);
-  void FinalCheck(std::vector<Instruction *> &illegalOffsets, Function &F,
-                  hlsl::OP *hlslOP);
+  void LegalizeOffsets(const std::vector<std::pair<Value *, CallInst*>> &illegalOffsets);
+  void FinalCheck(Function &F, hlsl::OP *hlslOP);
 };
 
 char DxilLegalizeSampleOffsetPass::ID = 0;
 
-bool HasIllegalOffsetInLoop(std::vector<Instruction *> &illegalOffsets,
+bool HasIllegalOffsetInLoop(std::vector<std::pair<Value *, CallInst*>> &illegalOffsets, LoopInfo &LI,
                             Function &F) {
   DominatorTreeAnalysis DTA;
   DominatorTree DT = DTA.run(F);
-  LoopInfo LI;
   LI.Analyze(DT);
 
   bool findOffset = false;
 
-  for (Instruction *I : illegalOffsets) {
-    BasicBlock *BB = I->getParent();
-    if (LI.getLoopFor(BB)) {
-      findOffset = true;
-      break;
+  for (auto it : illegalOffsets) {
+    if (Instruction *I = dyn_cast<Instruction>(it.first)) {
+      BasicBlock *BB = I->getParent();
+      if (LI.getLoopFor(BB)) {
+        findOffset = true;
```

Comment:
I didn't actually change any of the function of this code. @adam-yang might have thoughts on that suggestion. 

Nevertheless, I'm happy to add the TODO.

The error message makes clear that manual unrolling is a speculative fix.

Reply:
Yes, this was just something I thought of while reading the code that could be captured somewhere relevant, like in a TODO comment, rather than just forgetting about it again.  Of course, it only makes sense to add a comment when already updating the PR for a better reason (and there were other reasons to do so).